enterprise/server/remote_execution/runner/runner.go

```diff
@@ -398,23 +408,30 @@ func (p *Pool) add(ctx context.Context, r *CommandRunner) *labeledError {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	if p.isShuttingDown {
-		return &labeledError{
-			status.UnavailableError("pool is shutting down; cannot add new runners"),
-			"pool_shutting_down",
-		}
-	}
+	r.state = paused
+	// Push the runner to the end of the list to ensure that we always evict
+	// the LRU runner.
+	p.remove(r)
+	p.runners = append(p.runners, r)
 
-	if len(p.runners) == p.maxRunnerCount {
-		if len(p.runners) == 0 {
+	if p.pausedRunnerCount() > p.maxRunnerCount {
+		if p.maxRunnerCount == 0 {
 			return &labeledError{
 				status.InternalError("pool max runner count is 0; this should never happen"),
 				"max_runner_count_zero",
 			}
 		}
-		// Evict the first and oldest runner to make room for the new one.
-		r := p.runners[0]
-		p.runners = p.runners[1:]
+		// Evict the oldest (first) paused runner to make room for the new one.
+		evictIndex := -1
+		for i, r := range p.runners {
+			if r.state == paused {
+				evictIndex = i
+				break
+			}
+		}
+
+		r := p.runners[evictIndex]
```

Comment:
does this work if evictIndex is still -1?

Reply:
It should never be -1 because the conditional only executes if there's a paused runner in the pool -- added a sanity check.

Reply:
I don't see it -- can you point me at it?

Reply:
I added a comment and moved some pieces of logic around a bit to make it more obvious what's going on. Because we only evict when the number of paused runners exceeds the max configured paused runner count (which we assert to be >= 1), there should be at least 1 runner to evict.